hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/BlockContentsValidator.java

```diff
@@ -0,0 +1,123 @@
+ */
+
+package com.hedera.services.bdd.junit.support.validators.block;
+
+import static com.hedera.services.bdd.junit.hedera.utils.WorkingDirUtils.workingDirFor;
+
+import com.hedera.hapi.block.stream.Block;
+import com.hedera.hapi.block.stream.BlockItem;
+import com.hedera.services.bdd.junit.support.BlockStreamAccess;
+import com.hedera.services.bdd.junit.support.BlockStreamValidator;
+import com.hedera.services.bdd.spec.HapiSpec;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.nio.file.Paths;
+import java.util.List;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.junit.jupiter.api.Assertions;
+
+public class BlockContentsValidator implements BlockStreamValidator {
+    private static final Logger logger = LogManager.getLogger(BlockContentsValidator.class);
+
+    public static void main(String[] args) {
+        final var node0Dir = Paths.get("hedera-node/test-clients")
+                .resolve(workingDirFor(0, "hapi"))
+                .toAbsolutePath()
+                .normalize();
+        final var validator = new BlockContentsValidator();
+        final var blocks =
+                BlockStreamAccess.BLOCK_STREAM_ACCESS.readBlocks(node0Dir.resolve("data/block-streams/block-0.0.3"));
+        validator.validateBlocks(blocks);
+    }
+
+    public static final Factory FACTORY = new Factory() {
+        @NonNull
+        @Override
+        public BlockStreamValidator create(@NonNull final HapiSpec spec) {
+            return new BlockContentsValidator();
+        }
+    };
+
+    @Override
+    public void validateBlocks(@NonNull final List<Block> blocks) {
+        for (int i = 0; i < blocks.size(); i++) {
+            try {
+                validate(blocks.get(i), i == 0);
+            } catch (AssertionError err) {
+                logger.error("Error validating block {}", blocks.get(i));
+                throw err;
+            }
+        }
+    }
+
+    private static void validate(Block block, boolean isFirst) {
+        final var blockItems = block.items();
+
+        // A block SHALL start with a `header`.
+        if (!blockItems.getFirst().hasBlockHeader()) {
+            Assertions.fail("Block does not start with a block header");
+        }
+
+        // A block SHALL end with a `state_proof`.
+        if (!blockItems.getLast().hasBlockProof()) {
+            Assertions.fail("Block does not end with a block proof");
+        }
+
+        // For the first block the `block_header` might be followed by `event_header` or `state_changes`, due to state
+        // migration.
+        if (isFirst) {
+            if (!blockItems.get(1).hasEventHeader() && !blockItems.get(1).hasStateChanges()) {
+                Assertions.fail("First block header not followed by event header or state changes");
+            }
+            return;
+        }
+
+        // TODO: In general, a `block_header` SHALL be followed by an `event_header`, but for hapiTestRestart we get
```

Comment:
@tinker-michaelj can you advise if that's acceptable or should we enforce it here and apply a fix to the code that produces the blocks?

Reply:
Ah in that case the post-restart `State` initialization changes state _before_ any event has reached consensus. 

I think this is fine.

Reply:
Updated code comment.